File: .claude/worktrees/hardcore-shockley-fe2352/services/enrichment/workers/itunes.js
'use strict';
// ── iTunes/Apple Music worker ─────────────────────────────────────────────────
// API: https://itunes.apple.com/search  — geen API key nodig
// Rate limit: max 20 calls/minuut (conservatief: 1 call per 4 sec)

const ITUNES_BASE    = 'https://itunes.apple.com';
const RATE_INTERVAL  = 4000; // ms tussen calls

class iTunesWorker {
  constructor(db, log) {
    this.db  = db;
    this.log = log.child ? log.child({ worker: 'itunes' }) : log;
    this._lastCall = 0;
  }

  /** Wacht totdat de rate-limit voorbij is. */
  async _rateLimit() {
    const wait = Math.max(0, RATE_INTERVAL - (Date.now() - this._lastCall));
    if (wait > 0) await new Promise(r => setTimeout(r, wait));
    this._lastCall = Date.now();
  }

  /**
   * Verwerk één queue-item.
   * @param {{ entity_type, entity_name, entity_id }} entity
   * @returns {{ ok: boolean, data?: object, error?: string }}
   */
  async process(entity) {
    try {
      const data = await this.search(entity.entity_name, entity.entity_type);
      if (!data) return { ok: false, error: 'No results found' };
      return { ok: true, data };
    } catch (err) {
      this.log.warn({ err: err.message, entity: entity.entity_name }, 'iTunes fetch failed');
      return { ok: false, error: err.message };
    }
  }

  /**
   * Zoek een artiest/album/track op bij iTunes.
   * @param {string} name
   * @param {'artist'|'album'|'track'} entityType
   */
  async search(name, entityType = 'artist') {
    await this._rateLimit();

    const itunesEntity = entityType === 'artist'
      ? 'musicArtist'
      : entityType === 'album'
        ? 'album'
        : 'musicTrack';

    const url = `${ITUNES_BASE}/search?${new URLSearchParams({
      term:   name,
      entity: itunesEntity,
      limit:  '5',
      media:  'music',
    })}`;

    const res = await fetch(url, {
      headers: { 'Accept': 'application/json' },
      signal:  AbortSignal.timeout(10_000),
    });

    if (!res.ok) throw new Error(`iTunes HTTP ${res.status}`);

    const json = await res.json();
    const results = json.results || [];
    if (!results.length) return null;

    // Zoek naar exacte match
    const norm  = s => (s || '').toLowerCase().trim();
    const exact = results.find(r =>
      norm(r.artistName || r.collectionName || r.trackName) === norm(name)
    );
    const best = exact || results[0];

    if (entityType === 'artist') {
      return this._mapArtist(best, results);
    } else if (entityType === 'album') {
      return this._mapAlbum(best);
    }
    return this._mapTrack(best);
  }

  _mapArtist(item, allResults) {
    return {
      artistId:     item.artistId,
      artistName:   item.artistName,
      artistType:   item.artistType,
      primaryGenre: item.primaryGenreName,
      genres:       [...new Set(allResults.map(r => r.primaryGenreName).filter(Boolean))],
      artistLinkUrl: item.artistLinkUrl,
      artworkUrl:   item.artworkUrl100 || null,
      source:       'itunes',
      fetchedAt:    Date.now(),
    };
  }

  _mapAlbum(item) {
    return {
      collectionId:   item.collectionId,
      collectionName: item.collectionName,
      artistName:     item.artistName,
      artworkUrl:     (item.artworkUrl100 || '').replace('100x100', '600x600'),
      releaseDate:    item.releaseDate,
      trackCount:     item.trackCount,
      primaryGenre:   item.primaryGenreName,
      collectionPrice: item.collectionPrice,
      currency:       item.currency,
      country:        item.country,
      source:         'itunes',
      fetchedAt:      Date.now(),
    };
  }

  _mapTrack(item) {
    return {
      trackId:        item.trackId,
      trackName:      item.trackName,
      artistName:     item.artistName,
      collectionName: item.collectionName,
      previewUrl:     item.previewUrl,
      artworkUrl:     (item.artworkUrl100 || '').replace('100x100', '600x600'),
      trackPrice:     item.trackPrice,
      releaseDate:    item.releaseDate,
      primaryGenre:   item.primaryGenreName,
      trackTimeMillis: item.trackTimeMillis,
      source:         'itunes',
      fetchedAt:      Date.now(),
    };
  }
}

module.exports = { iTunesWorker };

File: core/workers/itunes.py
"""
itunes.py — iTunes/Apple Music enrichment worker.

Gebruikt de gratis iTunes Search API (geen auth nodig).
Rate limit: max ~20 calls/min → 1 call per 4 sec.
"""

import logging
import time
from typing import Any, Optional

import requests

from core.workers.base import BaseWorker

ITUNES_BASE   = "https://itunes.apple.com"
RATE_INTERVAL = 4.0   # seconden
REQUEST_TIMEOUT = 10


class ITunesWorker(BaseWorker):
    source = "itunes"

    def __init__(self, database, logger: logging.Logger):
        super().__init__(database, logger)
        self._last_call = 0.0
        self._session   = requests.Session()
        self._session.headers.update({"Accept": "application/json"})

    def _rate_limit(self) -> None:
        wait = max(0.0, RATE_INTERVAL - (time.time() - self._last_call))
        if wait > 0:
            time.sleep(wait)
        self._last_call = time.time()

    def process(self, item: dict) -> dict[str, Any]:
        try:
            data = self._search(item["entity_name"], item["entity_type"])
            if not data:
                return {"ok": False, "error": "No results found"}
            return {"ok": True, "data": data}
        except Exception as exc:
            self.log.warning(f"iTunes worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}

    def _search(self, name: str, entity_type: str = "artist") -> Optional[dict]:
        self._rate_limit()

        itunes_entity = (
            "musicArtist" if entity_type == "artist" else
            "album"       if entity_type == "album"  else
            "musicTrack"
        )

        resp = self._session.get(
            f"{ITUNES_BASE}/search",
            params={"term": name, "entity": itunes_entity, "limit": 5, "media": "music"},
            timeout=REQUEST_TIMEOUT,
        )
        resp.raise_for_status()

        results = resp.json().get("results", [])
        if not results:
            return None

        # Exacte match op naam-veld
        norm   = lambda s: (s or "").lower().strip()
        fields = ["artistName", "collectionName", "trackName"]
        exact  = next(
            (r for r in results
             if any(norm(r.get(f, "")) == norm(name) for f in fields)),
            None,
        )
        best = exact or results[0]

        if entity_type == "artist":
            return self._map_artist(best, results)
        if entity_type == "album":
            return self._map_album(best)
        return self._map_track(best)

    def _map_artist(self, item: dict, all_results: list) -> dict:
        genres = list({r["primaryGenreName"] for r in all_results if r.get("primaryGenreName")})
        return {
            "artistId":      item.get("artistId"),
            "artistName":    item.get("artistName"),
            "artistType":    item.get("artistType"),
            "primaryGenre":  item.get("primaryGenreName"),
            "genres":        genres,
            "artistLinkUrl": item.get("artistLinkUrl"),
            "artworkUrl":    item.get("artworkUrl100"),
            "source":        "itunes",
            "fetchedAt":     int(time.time() * 1000),
        }

    def _map_album(self, item: dict) -> dict:
        artwork = (item.get("artworkUrl100") or "").replace("100x100", "600x600")
        return {
            "collectionId":    item.get("collectionId"),
            "collectionName":  item.get("collectionName"),
            "artistName":      item.get("artistName"),
            "artworkUrl":      artwork or None,
            "releaseDate":     item.get("releaseDate"),
            "trackCount":      item.get("trackCount"),
            "primaryGenre":    item.get("primaryGenreName"),
            "collectionPrice": item.get("collectionPrice"),
            "currency":        item.get("currency"),
            "country":         item.get("country"),
            "source":          "itunes",
            "fetchedAt":       int(time.time() * 1000),
        }

    def _map_track(self, item: dict) -> dict:
        artwork = (item.get("artworkUrl100") or "").replace("100x100", "600x600")
        return {
            "trackId":         item.get("trackId"),
            "trackName":       item.get("trackName"),
            "artistName":      item.get("artistName"),
            "collectionName":  item.get("collectionName"),
            "previewUrl":      item.get("previewUrl"),
            "artworkUrl":      artwork or None,
            "trackPrice":      item.get("trackPrice"),
            "releaseDate":     item.get("releaseDate"),
            "primaryGenre":    item.get("primaryGenreName"),
            "trackTimeMillis": item.get("trackTimeMillis"),
            "source":          "itunes",
            "fetchedAt":       int(time.time() * 1000),
        }
